### PR TITLE
Fix Triplet Margin Loss Opinfo

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110302

Triplet Margin Loss takes in a Callable `distance_function` parameter which is not supported as an argument on the fx graph. See previous error:

> File "/scratch/eellison/work/pytorch/torch/_dynamo/symbolic_convert.py", line 562, in call_function
self.push(fn.call_function(self, args, kwargs))
File "/scratch/eellison/work/pytorch/torch/_dynamo/variables/torch.py", line 723, in call_function
*proxy_args_kwargs(args, kwargs),
File "/scratch/eellison/work/pytorch/torch/_dynamo/utils.py", line 504, in proxy_args_kwargs
f"call_function args: {typestr(*args)} {typestr(*list(kwargs.values()))}"
File "/scratch/eellison/work/pytorch/torch/_dynamo/exc.py", line 143, in unimplemented
raise Unsupported(msg)
torch._dynamo.exc.Unsupported: call_function args: TensorVariable() TensorVariable() TensorVariable() ConstantVariable(float) NNModuleVariable()

This is fixable by just inlining into `triplet_margin_loss` and continuing to compile it. This required support for `has_torch_function_variadic`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by int3** [comment]
> This required support for has_torch_function_variadic.

Could you explain a bit more about how `has_torch_function_variadic` comes in?

File: torch/_dynamo/variables/torch.py [hunk 1]
```diff
@@ -359,6 +359,12 @@ def call_function(
             )
             assert len(args) == 1
             return CUDAStreamContextVariable.create(tx, args[0], **options)
+        elif self.value in (torch.overrides.has_torch_function_variadic, torch.overrides.has_torch_function_unary):
+            assert not kwargs
+            def has_torch_function(a):
+                return isinstance(a, TensorWithTFOverrideVariable) or isinstance(a, UserDefinedObjectVariable) and hasattr(a.value, "__torch_function__")
```
Comment (Skylion007) [inline torch/_dynamo/variables/torch.py:365]: You probably want to add parantheses here as the order of operations isn't quite clear to a non-experience Python reader:
```suggestion
                return isinstance(a, TensorWithTFOverrideVariable) or (isinstance(a, UserDefinedObjectVariable) and hasattr(a.value, "__torch_function__"))
```
Or merge the isinstance checks if that wasn't your intent.

**Comment by eellison** [comment]
@int3 it's because we are no longer inserting `triplet_margin_loss_with_distance` into the fx graph and recursing into it instead, so we end up compiling the `has_torch_function_variadic` here: https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L4645

File: torch/_dynamo/variables/torch.py [hunk 2]
```diff
@@ -359,6 +359,12 @@ def call_function(
             )
             assert len(args) == 1
             return CUDAStreamContextVariable.create(tx, args[0], **options)
+        elif self.value in (torch.overrides.has_torch_function_variadic, torch.overrides.has_torch_function_unary):
+            assert not kwargs
+            def has_torch_function(a):
```
Comment (mlazos) [inline torch/_dynamo/variables/torch.py:364]: Can you put this in utils.py, I have an impl in one of my torch function PRs

**Review by mlazos — APPROVED** [review]

**Pull request closed without merging**